File: backend/app/models/task.py
import uuid
from datetime import datetime, date, time

from sqlalchemy import (
    String, Boolean, DateTime, Date, Time, Integer, Text, ForeignKey,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Task(Base):
    __tablename__ = "tasks"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    contact_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("contacts.id", ondelete="SET NULL")
    )
    deal_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("deals.id", ondelete="SET NULL")
    )
    assigned_to: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), index=True
    )
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL")
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    priority: Mapped[str] = mapped_column(String(20), nullable=False, default="medium")
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending")
    type: Mapped[str] = mapped_column(String(50), nullable=False, default="follow_up")
    due_date: Mapped[date | None] = mapped_column(Date)
    due_time: Mapped[time | None] = mapped_column(Time)
    duration_minutes: Mapped[int | None] = mapped_column(Integer)
    is_all_day: Mapped[bool] = mapped_column(Boolean, default=True)
    recurrence: Mapped[str] = mapped_column(String(20), default="none")
    external_calendar_id: Mapped[str | None] = mapped_column(String(255))
    completed_at: Mapped[datetime | None] = mapped_column(DateTime)
    completed_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL")
    )
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    # Relationships
    contact: Mapped["Contact | None"] = relationship(
        "Contact", back_populates="tasks", lazy="selectin"
    )
    deal: Mapped["Deal | None"] = relationship("Deal", lazy="selectin")
    assigned_user: Mapped["User | None"] = relationship(
        "User", lazy="selectin", foreign_keys=[assigned_to]
    )
    created_by_user: Mapped["User | None"] = relationship(
        "User", lazy="selectin", foreign_keys=[created_by]
    )
    completed_by_user: Mapped["User | None"] = relationship(
        "User", lazy="selectin", foreign_keys=[completed_by]
    )
    comments: Mapped[list["TaskComment"]] = relationship(
        "TaskComment", back_populates="task", lazy="selectin",
        order_by="TaskComment.created_at.desc()"
    )

    def __repr__(self) -> str:
        return f"<Task {self.title}>"


class TaskComment(Base):
    __tablename__ = "task_comments"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    task_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("tasks.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    content: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, nullable=False
    )

    # Relationships
    task: Mapped["Task"] = relationship(
        "Task", back_populates="comments", lazy="selectin"
    )
    user: Mapped["User"] = relationship("User", lazy="selectin")

    def __repr__(self) -> str:
        return f"<TaskComment task={self.task_id}>"

File: frontend/src/types/task.ts
export interface Task {
  id: string
  org_id: string
  contact_id: string | null
  deal_id: string | null
  assigned_to: string | null
  created_by: string
  title: string
  description: string
  priority: TaskPriority
  status: TaskStatus
  type: TaskType
  due_date: string
  due_time: string | null
  duration_minutes: number
  is_all_day: boolean
  recurrence: TaskRecurrence | null
  completed_at: string | null
  completed_by: string | null
  created_at: string
  updated_at: string
}

export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent'
export type TaskStatus = 'pending' | 'in_progress' | 'completed' | 'cancelled'
export type TaskType = 'call' | 'email' | 'meeting' | 'site_visit' | 'install' | 'follow_up' | 'other'
export type TaskRecurrence = 'daily' | 'weekly' | 'monthly' | 'none'

export const TASK_PRIORITY_LABELS: Record<TaskPriority, string> = {
  low: 'Low',
  medium: 'Medium',
  high: 'High',
  urgent: 'Urgent',
}

export const TASK_STATUS_LABELS: Record<TaskStatus, string> = {
  pending: 'Pending',
  in_progress: 'In Progress',
  completed: 'Completed',
  cancelled: 'Cancelled',
}

export const TASK_TYPE_LABELS: Record<TaskType, string> = {
  call: 'Call',
  email: 'Email',
  meeting: 'Meeting',
  site_visit: 'Site Visit',
  install: 'Install',
  follow_up: 'Follow Up',
  other: 'Other',
}

export const TASK_RECURRENCE_LABELS: Record<TaskRecurrence, string> = {
  daily: 'Daily',
  weekly: 'Weekly',
  monthly: 'Monthly',
  none: 'None',
}
